Record CUDA events for "follow-up" FutureNCCL inside markCompleted

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48790 Add support for async callbacks in ivalue::Future
* #49014 Drop FutureNCCL in favor of vanilla CUDAFuture
* #48789 Make CUDAFuture remember and restore current device in callback
* #48840 Remove DataPtr extractor from CUDAFuture
* #48788 Cache the DataPtrs in CUDAFuture
* #48506 Split out reusable CUDAFuture from FutureNCCL
* #48505 Merge common parts of FutureNCCL into at::ivalue::Future
* #48504 Split FutureNCCL's CUDA-specific parts from generic future logic
* #48502 Support wider range of types in FutureNCCL
* #48501 Don't store device indices separately on FutureNCCL
* #48500 Add multi-GPU support to FutureNCCL
* #48563 Fix FutureNCCL not recording dataptrs with caching alloc in wait()
* #48503 Fix FutureNCCL's completed() disagreeing with wait()
* **#48499 Record CUDA events for "follow-up" FutureNCCL inside markCompleted**
* #48498 Use fresh stream from pool for each FutureNCCL callback
* #48497 Make FutureNCCL record events in current stream
* #48496 Have FutureNCCL record streams w/ allocator in addCallback
* #48562 Add some safeguards to FutureNCCL
* #48495 Remove NCCL dependency from PythonFutureWrapper
* #48561 Avoid using FutureNCCL before it's ready

This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

We can merge and "hide" a whole bunch of CUDA-related logic if we store and record the CUDA events that correspond to the completion of a FutureNCCL when we call markCompleted (rather than splitting it between the constructor, the `then` method, and a wrapper around the callback).

A more concrete reason for this change is that soon I'll add support for multi-device, and in that case we can't necessarily know in advance which devices a value will be on until we get that value (and we don't want to record an event on all devices as then we might "over-synchronize").

To me, this also makes more conceptual sense: the moment when we store a value on the future, which is the "signal" that the future is now ready, should also be time at which we record the events needed to synchronize with that value. Though this may just be personal preference.

Differential Revision: [D25177557](https://our.internmc.facebook.com/intern/diff/D25177557/)

lw committed:
Record CUDA events for "follow-up" FutureNCCL inside markCompleted

This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

We can merge and "hide" a whole bunch of CUDA-related logic if we store and record the CUDA events that correspond to the completion of a FutureNCCL when we call markCompleted (rather than splitting it between the constructor, the `then` method, and a wrapper around the callback).

A more concrete reason for this change is that soon I'll add support for multi-device, and in that case we can't necessarily know in advance which devices a value will be on until we get that value (and we don't want to record an event on all devices as then we might "over-synchronize").

To me, this also makes more conceptual sense: the moment when we store a value on the future, which is the "signal" that the future is now ready, should also be time at which we record the events needed to synchronize with that value. Though this may just be personal preference.

Differential Revision: [D25177557](https://our.internmc.facebook.com/intern/diff/D25177557/)

[ghstack-poisoned]

mrshenli commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -259,6 +249,15 @@ class ProcessGroupNCCL : public ProcessGroup {
           "outputs or the return value of the callback.");
       value_ = std::move(value);
+
+      if (cudaEvents_ == nullptr) {
```
Should this be an assert? Do we expect `cudaEvents_` be set twice on the same future object?

lw replied:
After this PR, the second constructor of FutureNCCL leaves `cudaEvents_` unset. So this `if` here basically checks whether this instance was created with that second constructor. It is thus totally expected for this to happen, even without `markCompleted` being called twice.

I agree that it's not pretty, but let me assure you it's temporary. If you bear with me until #48506 you'll see that when we split FutureNCCL into CUDAFuture + a thin FutureNCCL subclass this will go away, because the two constructors will go one in each of these two classes, and thus these checks will automatically be replaced by inheritance.

lw replied:
Oh, I get it now: when the future is created through the first constructor then `markCompleted` never gets called. Sorry I had missed that. In that case I think that, yes, this could become an assert. (It would have to be turned back into an `if` in #48505 though as there the first constructor will just call `markCompleted` internally).

mrshenli commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -259,6 +249,15 @@ class ProcessGroupNCCL : public ProcessGroup {
+        cudaEvents_ = std::make_shared<std::vector<at::cuda::CUDAEvent>>(1);
+        // In case of chained then callback calls, cudaEvents
+        // records callback's stream.
+        (*cudaEvents_)[0].record(at::cuda::getDefaultCUDAStream(deviceIndex_));
```
ditto to the previous PR, this should be current stream instead?

lw replied:
Yes, indeed I got confused here, should be fixed now.

mrshenli commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -223,19 +223,9 @@ class ProcessGroupNCCL : public ProcessGroup {
+    explicit FutureNCCL(c10::DeviceIndex deviceIndex)
         : at::ivalue::Future(c10::ListType::create(c10::TensorType::get())),
-          deviceIndex_(deviceIndex),
-          cudaEvents_(std::move(cudaEvents)) {
```
this means `C++` code can no longer call `completed()` after constructing `FutureNCCL` using this ctor?

lw replied:
I see what you mean and, yes, you are right. Note however that the FutureNCCL class (even before I started reworking it) relied very strongly on the implicit assumption that instances would be marked complete immediately upon construction. Without that assumption, the behavior of FutureNCCL would be fundamentally wrong (e.g., it always calls the callbacks immediately inline). So far, I am still relying on that assumption (I'll relax it in #48505). Hence, this second constructor is only called by the `then()` method, which however then runs the callback inline and thus when it returns the future to the user it's already complete. Thus it's still correct (even for C++ users) to always return true from `completed()`.

If you'd like some additional safety here, one thing I can do is to make this second constructor private, so that we can really be sure that this will only be used in the `then()` method. However, we'd have to make it public again in #48505.

lw committed:
Update on "Record CUDA events for "follow-up" FutureNCCL inside markCompleted"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

We can merge and "hide" a whole bunch of CUDA-related logic if we store and record the CUDA events that correspond to the completion of a FutureNCCL when we call markCompleted (rather than splitting it between the constructor, the `then` method, and a wrapper around the callback).

A more concrete reason for this change is that soon I'll add support for multi-device, and in that case we can't necessarily know in advance which devices a value will be on until we get that value (and we don't want to record an event on all devices as then we might "over-synchronize").

To me, this also makes more conceptual sense: the moment when we store a value on the future, which is the "signal" that the future is now ready, should also be time at which we record the events needed to synchronize with that value. Though this may just be personal preference.

Differential Revision: [D25177557](https://our.internmc.facebook.com/intern/diff/D25177557/)

[ghstack-poisoned]

lw committed:
Update on "Record CUDA events for "follow-up" FutureNCCL inside markCompleted"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

We can merge and "hide" a whole bunch of CUDA-related logic if we store and record the CUDA events that correspond to the completion of a FutureNCCL when we call markCompleted (rather than splitting it between the constructor, the `then` method, and a wrapper around the callback).

A more concrete reason for this change is that soon I'll add support for multi-device, and in that case we can't necessarily know in advance which devices a value will be on until we get that value (and we don't want to record an event on all devices as then we might "over-synchronize").

To me, this also makes more conceptual sense: the moment when we store a value on the future, which is the "signal" that the future is now ready, should also be time at which we record the events needed to synchronize with that value. Though this may just be personal preference.

Differential Revision: [D25177557](https://our.internmc.facebook.com/intern/diff/D25177557/)

[ghstack-poisoned]

lw committed:
Update on "Record CUDA events for "follow-up" FutureNCCL inside markCompleted"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

We can merge and "hide" a whole bunch of CUDA-related logic if we store and record the CUDA events that correspond to the completion of a FutureNCCL when we call markCompleted (rather than splitting it between the constructor, the `then` method, and a wrapper around the callback).

A more concrete reason for this change is that soon I'll add support for multi-device, and in that case we can't necessarily know in advance which devices a value will be on until we get that value (and we don't want to record an event on all devices as then we might "over-synchronize").

To me, this also makes more conceptual sense: the moment when we store a value on the future, which is the "signal" that the future is now ready, should also be time at which we record the events needed to synchronize with that value. Though this may just be personal preference.

Differential Revision: [D25177557](https://our.internmc.facebook.com/intern/diff/D25177557/)

[ghstack-poisoned]

mrshenli reviewed (approved):
LGTM!

lw committed:
Update on "Record CUDA events for "follow-up" FutureNCCL inside markCompleted"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

We can merge and "hide" a whole bunch of CUDA-related logic if we store and record the CUDA events that correspond to the completion of a FutureNCCL when we call markCompleted (rather than splitting it between the constructor, the `then` method, and a wrapper around the callback).

A more concrete reason for this change is that soon I'll add support for multi-device, and in that case we can't necessarily know in advance which devices a value will be on until we get that value (and we don't want to record an event on all devices as then we might "over-synchronize").

To me, this also makes more conceptual sense: the moment when we store a value on the future, which is the "signal" that the future is now ready, should also be time at which we record the events needed to synchronize with that value. Though this may just be personal preference.

Differential Revision: [D25177557](https://our.internmc.facebook.com/intern/diff/D25177557/)

[ghstack-poisoned]

lw committed:
Update on "Record CUDA events for "follow-up" FutureNCCL inside markCompleted"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

We can merge and "hide" a whole bunch of CUDA-related logic if we store and record the CUDA events that correspond to the completion of a FutureNCCL when we call markCompleted (rather than splitting it between the constructor, the `then` method, and a wrapper around the callback).

A more concrete reason for this change is that soon I'll add support for multi-device, and in that case we can't necessarily know in advance which devices a value will be on until we get that value (and we don't want to record an event on all devices as then we might "over-synchronize").

To me, this also makes more conceptual sense: the moment when we store a value on the future, which is the "signal" that the future is now ready, should also be time at which we record the events needed to synchronize with that value. Though this may just be personal preference.

Differential Revision: [D25177557](https://our.internmc.facebook.com/intern/diff/D25177557/)

[ghstack-poisoned]

lw committed:
Update on "Record CUDA events for "follow-up" FutureNCCL inside markCompleted"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

We can merge and "hide" a whole bunch of CUDA-related logic if we store and record the CUDA events that correspond to the completion of a FutureNCCL when we call markCompleted (rather than splitting it between the constructor, the `then` method, and a wrapper around the callback).

A more concrete reason for this change is that soon I'll add support for multi-device, and in that case we can't necessarily know in advance which devices a value will be on until we get that value (and we don't want to record an event on all devices as then we might "over-synchronize").

To me, this also makes more conceptual sense: the moment when we store a value on the future, which is the "signal" that the future is now ready, should also be time at which we record the events needed to synchronize with that value. Though this may just be personal preference.

Differential Revision: [D25177557](https://our.internmc.facebook.com/intern/diff/D25177557/)

[ghstack-poisoned]

lw committed:
Update on "Record CUDA events for "follow-up" FutureNCCL inside markCompleted"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

We can merge and "hide" a whole bunch of CUDA-related logic if we store and record the CUDA events that correspond to the completion of a FutureNCCL when we call markCompleted (rather than splitting it between the constructor, the `then` method, and a wrapper around the callback).

A more concrete reason for this change is that soon I'll add support for multi-device, and in that case we can't necessarily know in advance which devices a value will be on until we get that value (and we don't want to record an event on all devices as then we might "over-synchronize").

To me, this also makes more conceptual sense: the moment when we store a value on the future, which is the "signal" that the future is now ready, should also be time at which we record the events needed to synchronize with that value. Though this may just be personal preference.

Differential Revision: [D25177557](https://our.internmc.facebook.com/intern/diff/D25177557/)

[ghstack-poisoned]
